### PR TITLE
New amqp sender service bean

### PR DESCRIPTION
**Pull request by SirWayne**
1. Add a AmqpSenderService interface, that the user can declare a own sender service to customize sending amqp messages.
The default implemenation use the the srping rabbit mq template to send messages.

2. Now the target adress saves the amqp reply to uri (vhost + exchange)

3. Now the current user

### CODE REVIEW

File: hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageDispatcherService.java [hunk 1]
```diff
@@ -25,34 +25,37 @@
...
-    private RabbitTemplate rabbitTemplate;
+    private ArtifactUrlHandler artifactUrlHandler;
 
     @Autowired
-    private TenantAware tenantAware;
+    private AmqpSenderService amqpSenderService;
 
     @Autowired
-    private ArtifactUrlHandler artifactUrlHandler;
+    public AmqpMessageDispatcherService(final MessageConverter messageConverter, final RabbitTemplate defaultTemplate) {
```
Comment (michahirsch) [inline hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageDispatcherService.java]: javadoc for constructor and parameters

File: hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java [hunk 2]
```diff
@@ -428,12 +409,12 @@ void setHostnameResolver(final HostnameResolver hostnameResolver) {
         this.hostnameResolver = hostnameResolver;
     }
 
-    void setRabbitTemplate(final RabbitTemplate rabbitTemplate) {
-        this.rabbitTemplate = rabbitTemplate;
+    void setMessageConverter(final MessageConverter messageConverter) {
```
Comment (michahirsch) [inline hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java]: Can this setter be removed? The `MessageConverter` is already given in the constructor and could be final I guess then.

File: hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/BaseAmqpService.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * A base class which provide basis amqp staff.
+ */
+public class BaseAmqpService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BaseAmqpService.class);
+    protected MessageConverter messageConverter;
+
+    protected RabbitTemplate internalAmqpTemplate;
+
+    public BaseAmqpService(final MessageConverter messageConverter, final RabbitTemplate defaultTemplate) {
```
Comment (michahirsch) [inline hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/BaseAmqpService.java]: Javadoc on constructor

File: hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/BaseAmqpService.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        this.messageConverter = messageConverter;
+        internalAmqpTemplate = defaultTemplate;
+    }
+
+    protected void cleanMessage(final Message message) {
```
Comment (michahirsch) [inline hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/BaseAmqpService.java]: protected methods should have javadoc due they can be overriden by subclasses

File: hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpSenderService.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+     * 
+     * @param message
+     *            the amqp message
+     * @param uri
+     *            the reply to uri
+     */
+    void sendMessage(Message message, URI uri);
+
+    /**
+     * Extract the exchange from the uri. Default implementation removes the
```
Comment (michahirsch) [inline hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpSenderService.java]: Add an exmaple URI pattern to the javadoc would be easier to know which kind of URI pattern need to be extracted. e.g amqp://<virtualHost>/<exchange>

File: hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java [hunk 6]
```diff
@@ -381,38 +373,12 @@ private void handleCancelRejected(final Message message, final Action action, fi
             // back to running action status
 
         } else {
-            logAndThrowMessageError(message, "Cancel Recjected message is not allowed, if action is on state: "
-                    + action.getStatus());
+            logAndThrowMessageError(message,
+                    "Cancel Recjected message is not allowed, if action is on state: " + action.getStatus());
```
Comment (Jonkno) [inline hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java]: rejected not Recjected

**Pull request merged**